File: executor-node/src/ipc/frame.ts
/**
 * IPC framing implementation per CONTRACT_IPC.md.
 *
 * Frame structure:
 * - 4-byte big-endian length prefix (unsigned, big-endian)
 * - payload bytes (msgpack-encoded)
 *
 * Constraints:
 * - Maximum frame size: 16 MiB (including length prefix)
 * - Maximum payload size: 16 MiB - 4 bytes
 * - Artifact chunk size: up to 8 MiB (raw bytes, before msgpack encoding)
 *
 * Two frame types exist, both msgpack-encoded:
 * - Event frame: msgpack-encoded EventEnvelope (discriminated by envelope.type)
 * - Artifact chunk frame: msgpack-encoded chunk envelope with type='artifact_chunk'
 *
 * Decoding discrimination: if decoded.type === 'artifact_chunk', it's a chunk
 * frame; otherwise it's an event envelope (type will be 'item', 'log', etc.).
 *
 * @module
 * @remarks Node.js only. Uses Buffer for transport efficiency.
 */

import { encode as msgpackEncode } from '@msgpack/msgpack'
import type { ArtifactId, EventEnvelope } from '@pithecene-io/quarry-sdk'

/**
 * Maximum frame size in bytes (16 MiB), including length prefix.
 * Frames exceeding this limit are invalid and must be rejected.
 */
export const MAX_FRAME_SIZE = 16 * 1024 * 1024

/**
 * Maximum payload size in bytes (MAX_FRAME_SIZE - LENGTH_PREFIX_SIZE).
 */
export const MAX_PAYLOAD_SIZE = MAX_FRAME_SIZE - 4

/**
 * Maximum artifact chunk size in bytes (8 MiB).
 * This is the limit on raw bytes before msgpack encoding.
 * Artifacts larger than this must be split into multiple chunks.
 */
export const MAX_CHUNK_SIZE = 8 * 1024 * 1024

/**
 * Length prefix size in bytes.
 */
export const LENGTH_PREFIX_SIZE = 4

/**
 * Frame type discriminant for artifact chunk frames.
 * Event frames use the EventEnvelope's own type field ('item', 'log', etc.).
 */
export type ArtifactChunkType = 'artifact_chunk'

/**
 * Frame type discriminant for run result control frames.
 */
export type RunResultType = 'run_result'

/**
 * Artifact chunk frame per CONTRACT_IPC.md.
 * This is a stream-level construct, not a normal emit event.
 *
 * @remarks
 * The `data` field is encoded as msgpack bin type (not array).
 * Receivers must decode it as raw bytes.
 */
export type ArtifactChunkFrame = {
  readonly type: 'artifact_chunk'
  readonly artifact_id: ArtifactId
  /** Sequence number, starts at 1 */
  readonly seq: number
  /** True if this is the last chunk */
  readonly is_last: boolean
  /** Raw binary data (msgpack bin type) */
  readonly data: Uint8Array
}

/**
 * Run result outcome status.
 */
export type RunResultStatus = 'completed' | 'error' | 'crash'

/**
 * Run result outcome per CONTRACT_IPC.md.
 * Describes the final outcome of a run.
 */
export type RunResultOutcome = {
  /** Outcome status */
  readonly status: RunResultStatus
  /** Human-readable message */
  readonly message?: string
  /** Error type (for error status) */
  readonly error_type?: string
  /** Stack trace (for error status) */
  readonly stack?: string
}

/**
 * Redacted proxy endpoint for run results (no password).
 * Per CONTRACT_PROXY.md: proxy_used must exclude password fields.
 */
export type ProxyEndpointRedactedFrame = {
  readonly protocol: 'http' | 'https' | 'socks5'
  readonly host: string
  readonly port: number
  readonly username?: string
}

/**
 * Run result control frame per CONTRACT_IPC.md.
 * This is a control frame emitted once after terminal event emission.
 * It does NOT affect seq ordering (not an event).
 */
export type RunResultFrame = {
  readonly type: 'run_result'
  /** Final run outcome */
  readonly outcome: RunResultOutcome
  /** Proxy endpoint used (redacted, no password) */
  readonly proxy_used?: ProxyEndpointRedactedFrame
}

/**
 * File write frame for sidecar file uploads via Lode Store.
 * Single-frame transport (not chunked). Max 8 MiB data.
 * Bypasses seq numbering, policy pipeline, and artifact manager.
 */
export type FileWriteFrame = {
  readonly type: 'file_write'
  /** Target filename (no path separators, no "..") */
  readonly filename: string
  /** MIME content type */
  readonly content_type: string
  /** Raw binary data (max MAX_CHUNK_SIZE) */
  readonly data: Uint8Array
}

/**
 * Union of all frame payload types for decoding.
 * Discriminate using type field:
 * - 'artifact_chunk' → ArtifactChunkFrame
 * - 'file_write' → FileWriteFrame (sidecar file upload)
 * - 'run_result' → RunResultFrame (control, not counted in seq)
 * - other (item, log, etc.) → EventEnvelope
 */
export type Frame = EventEnvelope | ArtifactChunkFrame | RunResultFrame | FileWriteFrame

/**
 * Error thrown when a frame exceeds the maximum size.
 */
export class FrameSizeError extends Error {
  constructor(
    public readonly payloadSize: number,
    public readonly maxPayloadSize: number
  ) {
    super(`Payload size ${payloadSize} exceeds maximum ${maxPayloadSize}`)
    this.name = 'FrameSizeError'
  }
}

/**
 * Encode data into a framed buffer with length prefix.
 *
 * @param payload - The payload bytes to frame
 * @returns Buffer containing length prefix + payload
 * @throws FrameSizeError if payload exceeds MAX_PAYLOAD_SIZE
 *
 * @remarks
 * The total frame size (prefix + payload) is bounded by MAX_FRAME_SIZE.
 */
export function encodeFrame(payload: Uint8Array): Buffer {
  if (payload.length > MAX_PAYLOAD_SIZE) {
    throw new FrameSizeError(payload.length, MAX_PAYLOAD_SIZE)
  }

  const frame = Buffer.allocUnsafe(LENGTH_PREFIX_SIZE + payload.length)
  frame.writeUInt32BE(payload.length, 0)
  frame.set(payload, LENGTH_PREFIX_SIZE)

  return frame
}

/**
 * Encode an event envelope into a framed buffer.
 * Per CONTRACT_IPC.md, the payload is the msgpack-encoded envelope directly.
 *
 * @param envelope - The event envelope to encode
 * @returns Buffer containing length prefix + msgpack-encoded envelope
 * @throws FrameSizeError if encoded payload exceeds MAX_PAYLOAD_SIZE
 */
export function encodeEventFrame(envelope: EventEnvelope): Buffer {
  const payload = msgpackEncode(envelope)
  return encodeFrame(payload)
}

/**
 * Error thrown when chunk parameters violate CONTRACT_IPC constraints.
 */
export class ChunkValidationError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ChunkValidationError'
  }
}

/**
 * Encode an artifact chunk into a framed buffer.
 *
 * @param artifactId - The artifact ID
 * @param seq - Sequence number (must be >= 1 per CONTRACT_IPC)
 * @param isLast - True if this is the final chunk
 * @param data - Raw binary data (must be <= MAX_CHUNK_SIZE per CONTRACT_IPC)
 * @returns Buffer containing length prefix + msgpack-encoded frame
 * @throws ChunkValidationError if seq < 1 or data exceeds MAX_CHUNK_SIZE
 * @throws FrameSizeError if encoded payload exceeds MAX_PAYLOAD_SIZE
 */
export function encodeArtifactChunkFrame(
  artifactId: ArtifactId,
  seq: number,
  isLast: boolean,
  data: Uint8Array
): Buffer {
  // Validate per CONTRACT_IPC constraints
  if (seq < 1) {
    throw new ChunkValidationError(`seq must be >= 1, got ${seq}`)
  }
  if (data.length > MAX_CHUNK_SIZE) {
    throw new ChunkValidationError(
      `data size ${data.length} exceeds MAX_CHUNK_SIZE ${MAX_CHUNK_SIZE}`
    )
  }

  const frame: ArtifactChunkFrame = {
    type: 'artifact_chunk',
    artifact_id: artifactId,
    seq,
    is_last: isLast,
    data
  }
  const payload = msgpackEncode(frame)
  return encodeFrame(payload)
}

/**
 * Metadata for a single artifact chunk (without data).
 * Used by chunk iterators to avoid data copying.
 */
export type ChunkMeta = {
  readonly seq: number
  readonly isLast: boolean
  readonly offset: number
  readonly length: number
}

/**
 * Calculate chunk metadata for artifact data.
 * Does not copy data; returns offsets for slicing.
 *
 * @param totalSize - Total size of the artifact data
 * @returns Array of chunk metadata
 */
export function calculateChunks(totalSize: number): ChunkMeta[] {
  const chunks: ChunkMeta[] = []

  if (totalSize === 0) {
    // Empty artifact: single chunk with is_last=true
    chunks.push({ seq: 1, isLast: true, offset: 0, length: 0 })
    return chunks
  }

  let offset = 0
  let seq = 1

  while (offset < totalSize) {
    const remaining = totalSize - offset
    const length = Math.min(remaining, MAX_CHUNK_SIZE)
    const isLast = offset + length >= totalSize

    chunks.push({ seq, isLast, offset, length })

    offset += length
    seq++
  }

  return chunks
}

/**
 * Generator that yields encoded artifact chunk frames.
 * Memory-efficient: encodes one chunk at a time.
 *
 * @param artifactId - The artifact ID for the chunks
 * @param data - The binary data to chunk
 * @yields Encoded frame buffers ready for transport
 */
export function* encodeArtifactChunks(
  artifactId: ArtifactId,
  data: Buffer | Uint8Array
): Generator<Buffer, void, unknown> {
  const chunks = calculateChunks(data.length)

  for (const chunk of chunks) {
    const chunkData = data.subarray(chunk.offset, chunk.offset + chunk.length)
    yield encodeArtifactChunkFrame(artifactId, chunk.seq, chunk.isLast, chunkData)
  }
}

/**
 * Encode a run result control frame.
 *
 * Per CONTRACT_IPC.md, this is a control frame that:
 * - Is emitted once after terminal event emission attempt
 * - Does NOT affect seq ordering (not counted as an event)
 * - Contains outcome and optional proxy_used (redacted)
 *
 * @param outcome - The run outcome
 * @param proxyUsed - Optional redacted proxy endpoint (no password)
 * @returns Buffer containing length prefix + msgpack-encoded frame
 * @throws FrameSizeError if encoded payload exceeds MAX_PAYLOAD_SIZE
 */
export function encodeRunResultFrame(
  outcome: RunResultOutcome,
  proxyUsed?: ProxyEndpointRedactedFrame
): Buffer {
  const frame: RunResultFrame = {
    type: 'run_result',
    outcome,
    ...(proxyUsed && { proxy_used: proxyUsed })
  }
  const payload = msgpackEncode(frame)
  return encodeFrame(payload)
}

/**
 * Encode a file write frame for sidecar file uploads.
 *
 * File writes bypass seq numbering and the policy pipeline.
 * Max data size is MAX_CHUNK_SIZE (8 MiB).
 *
 * @param filename - Target filename (no path separators, no "..")
 * @param contentType - MIME content type
 * @param data - Raw binary data
 * @returns Buffer containing length prefix + msgpack-encoded frame
 * @throws ChunkValidationError if data exceeds MAX_CHUNK_SIZE
 * @throws FrameSizeError if encoded payload exceeds MAX_PAYLOAD_SIZE
 */
export function encodeFileWriteFrame(
  filename: string,
  contentType: string,
  data: Buffer | Uint8Array
): Buffer {
  if (data.length > MAX_CHUNK_SIZE) {
    throw new ChunkValidationError(
      `file data size ${data.length} exceeds MAX_CHUNK_SIZE ${MAX_CHUNK_SIZE}`
    )
  }

  const frame: FileWriteFrame = {
    type: 'file_write',
    filename,
    content_type: contentType,
    data
  }
  const payload = msgpackEncode(frame)
  return encodeFrame(payload)
}

File: quarry/ipc/frame.go
// Package ipc implements IPC framing per CONTRACT_IPC.md.
package ipc

import (
	"bufio"
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"io"

	"github.com/vmihailenco/msgpack/v5"

	"github.com/pithecene-io/quarry/types"
)

// Frame size constants per CONTRACT_IPC.md.
const (
	// MaxFrameSize is the maximum frame size (16 MiB), including length prefix.
	MaxFrameSize = 16 * 1024 * 1024
	// MaxPayloadSize is the maximum payload size (MaxFrameSize - 4 bytes).
	MaxPayloadSize = MaxFrameSize - LengthPrefixSize
	// MaxChunkSize is the maximum artifact chunk size (8 MiB raw bytes).
	MaxChunkSize = 8 * 1024 * 1024
	// LengthPrefixSize is the size of the length prefix in bytes.
	LengthPrefixSize = 4
)

// ArtifactChunkType is the type discriminant for artifact chunk frames.
const ArtifactChunkType = "artifact_chunk"

// RunResultType is the type discriminant for run result control frames.
const RunResultType = "run_result"

// FileWriteType is the type discriminant for file write frames.
// File writes bypass seq numbering and the policy pipeline.
const FileWriteType = "file_write"

// FileWriteAckType is the type discriminant for file write acknowledgement frames.
// Sent runtime→executor via stdin after processing a file_write frame.
const FileWriteAckType = "file_write_ack"

// FrameErrorKind classifies frame decoding errors.
type FrameErrorKind int

const (
	// FrameErrorPartial indicates a truncated or incomplete frame.
	FrameErrorPartial FrameErrorKind = iota
	// FrameErrorTooLarge indicates a frame exceeding MaxFrameSize.
	FrameErrorTooLarge
	// FrameErrorDecode indicates a msgpack decoding error.
	FrameErrorDecode
)

// FrameError represents a frame decoding error.
type FrameError struct {
	Kind FrameErrorKind
	Msg  string
	Err  error
}

func (e *FrameError) Error() string {
	if e.Err != nil {
		return fmt.Sprintf("%s: %v", e.Msg, e.Err)
	}
	return e.Msg
}

func (e *FrameError) Unwrap() error {
	return e.Err
}

// IsFatal returns true if this error is fatal (terminate run).
// Per CONTRACT_IPC.md, partial and oversized frames are fatal.
func (e *FrameError) IsFatal() bool {
	return e.Kind == FrameErrorPartial || e.Kind == FrameErrorTooLarge
}

// IsFatalFrameError returns true if the error is a fatal frame error.
func IsFatalFrameError(err error) bool {
	var frameErr *FrameError
	if errors.As(err, &frameErr) {
		return frameErr.IsFatal()
	}
	return false
}

// FrameDecoder decodes length-prefixed msgpack frames from a stream.
type FrameDecoder struct {
	reader io.Reader
}

// NewFrameDecoder creates a new frame decoder.
// Wraps the reader with bufio.Reader to reduce syscall overhead
// on unbuffered sources (e.g., OS pipes from child processes).
func NewFrameDecoder(r io.Reader) *FrameDecoder {
	br, ok := r.(*bufio.Reader)
	if !ok {
		br = bufio.NewReader(r)
	}
	return &FrameDecoder{reader: br}
}

// ReadFrame reads a single frame from the stream.
// Returns the raw payload bytes (msgpack-encoded).
//
// Errors:
//   - io.EOF: stream ended cleanly (no more frames)
//   - *FrameError with Kind=FrameErrorPartial: incomplete frame (fatal)
//   - *FrameError with Kind=FrameErrorTooLarge: frame exceeds limit (fatal)
func (d *FrameDecoder) ReadFrame() ([]byte, error) {
	// Read 4-byte big-endian length prefix
	var lengthBuf [LengthPrefixSize]byte
	_, err := io.ReadFull(d.reader, lengthBuf[:])
	if err != nil {
		if err == io.EOF {
			return nil, io.EOF
		}
		// Partial read of length prefix
		return nil, &FrameError{
			Kind: FrameErrorPartial,
			Msg:  "failed to read length prefix",
			Err:  err,
		}
	}

	payloadSize := binary.BigEndian.Uint32(lengthBuf[:])

	// Validate frame size per CONTRACT_IPC.md
	if payloadSize > MaxPayloadSize {
		return nil, &FrameError{
			Kind: FrameErrorTooLarge,
			Msg:  fmt.Sprintf("payload size %d exceeds maximum %d", payloadSize, MaxPayloadSize),
		}
	}

	// Read payload
	payload := make([]byte, payloadSize)
	_, err = io.ReadFull(d.reader, payload)
	if err != nil {
		return nil, &FrameError{
			Kind: FrameErrorPartial,
			Msg:  "failed to read payload",
			Err:  err,
		}
	}

	return payload, nil
}

// probeFrameType extracts the "type" field from a msgpack map without
// fully unmarshaling the payload. Falls back to full probe on error.
func probeFrameType(payload []byte) (string, error) {
	dec := msgpack.NewDecoder(bytes.NewReader(payload))
	n, err := dec.DecodeMapLen()
	if err != nil {
		return "", err
	}
	for range n {
		key, err := dec.DecodeString()
		if err != nil {
			return "", err
		}
		if key == "type" {
			return dec.DecodeString()
		}
		if err := dec.Skip(); err != nil {
			return "", err
		}
	}
	return "", errors.New("missing type field")
}

// DecodeFrame decodes a payload and returns a typed frame.
// Discriminates based on the type field: "artifact_chunk", "run_result",
// "file_write", or event types.
func DecodeFrame(payload []byte) (any, error) {
	frameType, err := probeFrameType(payload)
	if err != nil {
		return nil, &FrameError{
			Kind: FrameErrorDecode,
			Msg:  "failed to decode frame type",
			Err:  err,
		}
	}

	switch frameType {
	case ArtifactChunkType:
		return DecodeArtifactChunk(payload)
	case RunResultType:
		return DecodeRunResult(payload)
	case FileWriteType:
		return DecodeFileWrite(payload)
	case FileWriteAckType:
		return DecodeFileWriteAck(payload)
	default:
		return DecodeEventEnvelope(payload)
	}
}

// DecodeEventEnvelope decodes a payload as an EventEnvelope.
func DecodeEventEnvelope(payload []byte) (*types.EventEnvelope, error) {
	var envelope types.EventEnvelope
	if err := msgpack.Unmarshal(payload, &envelope); err != nil {
		return nil, &FrameError{
			Kind: FrameErrorDecode,
			Msg:  "failed to decode event envelope",
			Err:  err,
		}
	}
	return &envelope, nil
}

// DecodeArtifactChunk decodes a payload as an ArtifactChunkFrame.
func DecodeArtifactChunk(payload []byte) (*types.ArtifactChunkFrame, error) {
	var chunk types.ArtifactChunkFrame
	if err := msgpack.Unmarshal(payload, &chunk); err != nil {
		return nil, &FrameError{
			Kind: FrameErrorDecode,
			Msg:  "failed to decode artifact chunk",
			Err:  err,
		}
	}
	return &chunk, nil
}

// DecodeRunResult decodes a payload as a RunResultFrame.
func DecodeRunResult(payload []byte) (*types.RunResultFrame, error) {
	var result types.RunResultFrame
	if err := msgpack.Unmarshal(payload, &result); err != nil {
		return nil, &FrameError{
			Kind: FrameErrorDecode,
			Msg:  "failed to decode run result",
			Err:  err,
		}
	}
	return &result, nil
}

// DecodeFileWrite decodes a payload as a FileWriteFrame.
func DecodeFileWrite(payload []byte) (*types.FileWriteFrame, error) {
	var frame types.FileWriteFrame
	if err := msgpack.Unmarshal(payload, &frame); err != nil {
		return nil, &FrameError{
			Kind: FrameErrorDecode,
			Msg:  "failed to decode file write",
			Err:  err,
		}
	}
	return &frame, nil
}

// DecodeFileWriteAck decodes a payload as a FileWriteAckFrame.
func DecodeFileWriteAck(payload []byte) (*types.FileWriteAckFrame, error) {
	var frame types.FileWriteAckFrame
	if err := msgpack.Unmarshal(payload, &frame); err != nil {
		return nil, &FrameError{
			Kind: FrameErrorDecode,
			Msg:  "failed to decode file write ack",
			Err:  err,
		}
	}
	return &frame, nil
}

// EncodeFrame encodes a payload with a 4-byte big-endian length prefix.
// This is the public encoder counterpart to FrameDecoder.ReadFrame.
func EncodeFrame(payload []byte) []byte {
	buf := make([]byte, LengthPrefixSize+len(payload))
	binary.BigEndian.PutUint32(buf[:LengthPrefixSize], uint32(len(payload)))
	copy(buf[LengthPrefixSize:], payload)
	return buf
}

// EncodeFileWriteAck encodes a FileWriteAckFrame as a length-prefixed msgpack frame.
func EncodeFileWriteAck(ack *types.FileWriteAckFrame) ([]byte, error) {
	payload, err := msgpack.Marshal(ack)
	if err != nil {
		return nil, fmt.Errorf("failed to encode file write ack: %w", err)
	}
	return EncodeFrame(payload), nil
}
